akka-actor-typed/src/main/scala/akka/actor/typed/javadsl/ReceiveBuilder.scala

```diff
@@ -190,30 +186,39 @@ object ReceiveBuilder {
 }
 
 /**
- * Receive type for [[AbstractBehavior]]
- *
  * INTERNAL API
  */
 @InternalApi
-private final class BuiltReceive[T](
-    messageHandlers: List[ReceiveBuilder.Case[T, T]],
-    signalHandlers: List[ReceiveBuilder.Case[T, Signal]])
-    extends Receive[T] {
+private abstract class ArrayBasedBuiltReceive[T] extends Receive[T] {
   import ReceiveBuilder.Case
 
-  override def receiveMessage(msg: T): Behavior[T] = receive[T](msg, messageHandlers)
-
-  override def receiveSignal(msg: Signal): Behavior[T] = receive[Signal](msg, signalHandlers)
-
   @tailrec
-  private def receive[M](msg: M, handlers: List[Case[T, M]]): Behavior[T] =
-    handlers match {
-      case Case(cls, predicate, handler) :: tail =>
-        if ((cls.isEmpty || cls.get.isAssignableFrom(msg.getClass)) && (predicate.isEmpty || predicate.get.test(msg)))
-          handler(msg)
-        else receive[M](msg, tail)
-      case _ =>
-        Behaviors.unhandled
+  protected final def receive[M](msg: M, handlers: Array[Case[T, M]], idx: Int): Behavior[T] = {
+    if (handlers.length == 0) {
+      Behaviors.unhandled[T]
```

Comment:
Cleaner to check this in `build()` and just return `Behavior.empty` instead of ever coming when there are no handlers?